File: 3-matmult/matmult-mpi.c
/* File: matmult-mpi.c
 *
 * Purpose: 
 * 
 * Input:
 * 
 * Output:
 * 
 * Compile: mpicc -o matmult-mpi-c.o matmult-mpi.c
 * 
 * Run: mpiexec -np 10 ./matmult-mpi-c.o
 * 
 * Algorithm:
 * 
 * Note:
 * 
 * */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "mpi.h"

void FillMatrix(float *array, int size);
void PrintArray(float *array, int row, int column, char name[]);
void Multiply(float *matrixA, float *matrixB, float *matrixC, int sizeX, int sizeY);
float MultiplyArray(float *matrixA, float *matrixB, int size, int x, int y);
void Init(int *waste, int *n, int size, int *sum, int proccess);

int main(int argc, char *argv[])
{
    int rank, proccess, size = 2;
    int i, j, k, waste, n, processSize;
    float matrixA[size * size], matrixB[size * size], matrixC[size * size];

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &proccess);

    Init(&waste, &n, size, &i, proccess);

    if (rank == 0)
    {
        FillMatrix(matrixA, size);
        FillMatrix(matrixB, size);
    }

    MPI_Bcast(matrixA, size * size, MPI_FLOAT, 0, MPI_COMM_WORLD);
    MPI_Bcast(matrixB, size * size, MPI_FLOAT, 0, MPI_COMM_WORLD);

    if (rank == 0)
    {
        processSize = n;
        for (i = 1; i <= (proccess - 1); i++)
        {
            if (i == proccess - 1)
                processSize = waste + n;
            MPI_Recv(matrixC + ((i - 1) * size * n), processSize * size, MPI_FLOAT, i, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        }
        PrintArray(matrixA, size, size, "matrixA");
        PrintArray(matrixB, size, size, "matrixB");
        PrintArray(matrixC, size, size, "matrixC");
    }
    else
    {
        if (rank == proccess - 1)
            processSize = waste + n;
        else
            processSize = n;
        Multiply(matrixA, matrixB, matrixC, processSize, size);
        MPI_Send(matrixC, processSize * size, MPI_FLOAT, 0, 0, MPI_COMM_WORLD);
    }

    MPI_Finalize();
    return 0;
}

/**
* Llena una matriz cuadrada con valores aleatorios.
* array - Matriz cuadrada.
* size - Tamaño de la matriz.
*/
void FillMatrix(float *array, int size)
{
    int i, j;

    for (i = 0; i < size; i++)
        for (j = 0; j < size; j++)
            *(array + (i * size) + j) = rand() % (11 * (i + 1)) * 1.12;
}

/**
* Imprime una matriz con un mensaje y un identificador.
* array - Matriz.
* row - Numero de renglones de la matriz.
* column - Numero de columnas de la matriz.
* name - Mensaje para imprimir.
* id - Identificador del proceso.
*/
void PrintArray(float *array, int row, int column, char name[])
{
    int i, j;

    printf("%s\n", name);

    for (i = 0; i < row; i++)
    {
        for (j = 0; j < column; j++)
            printf("%.2f\t", *(array + (i * column) + j));
        printf("\n");
    }
}

/**
* Multiplica dos matrices cuadradas A, B y almacena el resultado en una tercera matriz C.
* matrixA - Primera matriz cuadrada.
* matrixB - Segunda matriz cuadrada.
* matrixC - Matriz resultado.
* size - Tamaño de las matrices.
*/
void Multiply(float *matrixA, float *matrixB, float *matrixC, int sizeX, int sizeY)
{
    int i, j;

    for (i = 0; i < sizeX; i++)
        for (j = 0; j < sizeY; j++)
            *(matrixC + (i * sizeX) + j) = MultiplyArray(matrixA, matrixB, sizeY, i, j);
}

/**
* Multiplica el renglon x de la matriz A por la columna y de la matriz B 
* y almacena el resultado una variable de regreso.
* matrixA - Primera matriz.
* matrixB - Segunda matriz.
* size - Tamaño de los arreglos.
* x - Renglon seleccionado de la matriz A. 
* y - Columna seleccionada de la matriz B.
* Forma de acceder a las posiciones de la matriz: matriz + renglon * tamaño + columna
*/
float MultiplyArray(float *matrixA, float *matrixB, int size, int x, int y)
{
    int i;
    float r = 0.0;

    for (i = 0; i < size; i++)
        r += *(matrixA + (x * size) + i) * *(matrixB + (i * size) + y);

    return r;
}

/**
* Inicializa el tamaño de informacion que debe repartir el primer proceso
* con todos los demas.
* waste - Residuo para el ultimo proceso.
* n -Tamaño de la informacion que se debe repartir entre los procesos.
* size -Tamaño de la informacion total.
* sum -Variable entera para inicializar en 0.
* process - Numero de procesos.
*/
void Init(int *waste, int *n, int size, int *sum, int proccess)
{
    *waste = size % (proccess - 1);
    *n = (size - *waste) / (proccess - 1);
    *sum = 0;
}

File: 3-matmult/matmult-mpi.f90
! File: matmult-mpi.f90
!
! Purpose: 
! 
! Input:
! 
! Output:
! 
! Compile: mpif90 -o matmult-mpi-f90.o matmult-mpi.f90
! 
! Run: mpiexec -np 10 ./matmult-mpi-f90.o
! 
! Algorithm:
! 
! Note:

program matmult
    implicit none
    include 'mpif.h'

    INTERFACE 
        subroutine PrintMatrix(array, size)
            REAL, INTENT(IN) :: array(:)
            INTEGER, INTENT(IN) :: size
        end subroutine PrintMatrix

        subroutine FillMatrix(array, size)
            REAL, INTENT(OUT) :: array(:)
            INTEGER, INTENT(IN) :: size
        end subroutine FillMatrix

        subroutine Multiply(MatrixA, MatrixB, MatrixC, sizeX, sizeY)
            REAL, INTENT(IN) :: MatrixA(:)
            REAL, INTENT(IN) :: MatrixB(:)
            REAL, INTENT(OUT) :: MatrixC(:)
            INTEGER, INTENT(IN) :: sizeX
            INTEGER, INTENT(IN) :: sizeY
        end subroutine Multiply

    END INTERFACE

    INTEGER, PARAMETER :: size = 3
    REAL, DIMENSION(size * size) :: matrixA
    REAL, DIMENSION(size * size) :: matrixB
    REAL, DIMENSION(size * size) :: matrixC
    INTEGER :: rank 
    INTEGER :: process
    INTEGER :: processSize
    INTEGER :: ierror
    INTEGER :: status(MPI_STATUS_SIZE)
    INTEGER :: waste
    INTEGER :: n
    INTEGER :: i

    call MPI_INIT(ierror)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, process, ierror)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)

    waste = mod(size, process - 1)
    n = size / (process - 1)

    if (rank == 0) then

        call FillMatrix(matrixA, size)
        call FillMatrix(matrixB, size)

    end if

    call MPI_BCAST(matrixA, size * size, MPI_REAL, 0, MPI_COMM_WORLD, ierror)
    call MPI_BCAST(matrixB, size * size, MPI_REAL, 0, MPI_COMM_WORLD, ierror)
    write (*, *) process, rank, n

    if (rank == 0) then
        
        processSize = n

        do i = 1, process - 1, 1

            if (i == process - 1) then
                processSize = waste + n
            end if

            call MPI_RECV(matrixC + ((i - 1) * size * n), processSize * size, MPI_REAL, i, 0, MPI_COMM_WORLD, ierror)

        end do
        call PrintMatrix(matrixA, size)
        call PrintMatrix(matrixB, size)
        call PrintMatrix(matrixC, size)

    else 

        if (rank == process - 1) then
            processSize = waste + n;
        else
            processSize = n;
        end if

        call Multiply(matrixA, matrixB, matrixC, processSize, size)
        call MPI_SEND(matrixC, processSize * size, MPI_REAL, 0, 0, MPI_COMM_WORLD)

    end if

    call MPI_FINALIZE(ierror)

end program matmult

subroutine PrintMatrix(array, size)
    IMPLICIT NONE
    REAL, INTENT(IN) :: array(:)
    INTEGER, INTENT(IN) :: size
    INTEGER :: i, j

    do i = 0, size - 1, 1
        do j = 0, size - 1, 1
            write (*,*) i, j, array(i * size + j) 
        end do
    end do

end subroutine PrintMatrix

subroutine FillMatrix(array, size)
    IMPLICIT NONE
    REAL, INTENT(OUT) :: array(:)
    INTEGER, INTENT(IN) :: size
    INTEGER :: i, j

    do i = 0, size - 1, 1
        do j = 0, size - 1, 1
            array(i * size + j)  = rand(i)
        end do
    end do

end subroutine FillMatrix

subroutine Multiply(MatrixA, MatrixB, MatrixC, sizeX, sizeY)
    IMPLICIT NONE
    REAL, INTENT(IN) :: MatrixA(:)
    REAL, INTENT(IN) :: MatrixB(:)
    REAL, INTENT(OUT) :: MatrixC(:)
    INTEGER, INTENT(IN) :: sizeX
    INTEGER, INTENT(IN) :: sizeY
    INTEGER :: i, j, k
    REAL :: result

    do i = 0, sizeX - 1, 1
        do j = 0, sizeY - 1, 1

            result = 0.0
            do k = 0, sizeY - 1, 1
                result = result + matrixA(i * sizeY + k) * matrixB(k * sizeY + j)
            end do
            matrixC(i * sizeY + j) = result
        
        end do
    end do

end subroutine Multiply